### DISCUSSION TITLE
Best way to integrate a react application within vscode webviews

### DISCUSSION THREAD

**Discussion by HatemMn** [microsoft/vscode-discussions#723] [Extension Development QnA] [answered]
Greetings,
I want to create an notebook extension where the sideBar is a webview coded in react, and interacts with the notebook renderer ( also coded in react).

There are two main issues with this : 

- What are the recommended ways of implementing react components within a [WebviewView](https://code.visualstudio.com/api/references/vscode-api#WebviewView) ? So far, I have a webpack config that compiles my react code into a file ( say dist/index.js ), and then within my webViewView html I read that script : `<script nonce="${nonce}" src="${webviewUris["dist/sideBarReact.js"]}"></script>`
Is this the best way ?
- Is it possible to wrap the components within the sideBar and those within the Notebook renderer in a common provider that will make them share the same state ?

**Comment by r3m0t** [comment, marked as answer]
1. This is correct, it's also possible to set up React Hot Reload, but not easy. To do that you use `webpack serve` and add a port mapping to the webview.
2. No as these are separate iframes. They both have access to webview messaging so you can send a message to the extension host when the state is updated and it can send a message to all webviews with the new state. The message could be the full state or just a reducer action. It's probably better to keep it simple and send the full state to avoid a bug causing the state to drift.

**Comment by HatemMn** [comment]
Thank you for the reply, for the second point, I do not understand what you meant by "The message could be the full state or just a reducer action." - is there a way in React to wrap the full state or tree of the application and send it to another application ?
To keep things simple, I am considering to create one component that will contain the state that shall be shared between the two iFrames, let's call it SharedParent, then have it wrap all the components of both apps. Am I supposed to use the Message Passing method for every single variable inside that component ?

**Reply by r3m0t** [reply]
Hi, it's a React question, and no there is no generic way to synchronise. I put all the state in the parent component which handles receiving messages from vscode. The lower components just receive props.

https://react.dev/learn/sharing-state-between-components

**Reply by HatemMn** [reply]
Thank you very much for the help !
Can @isidorn please the other reply as the selected answer ? It replied to my first question
